src/foam/nanos/notification/email/DAOEmailService.js

```diff
@@ -92,55 +112,173 @@ return config_;`
+        if ( emailMessage == null )
+          return;
 
-EnvironmentConfiguration config = getConfig(group);
-JtwigModel model = JtwigModel.newModel(templateArgs);
-emailMessage = (EmailMessage) emailMessage.fclone();
+        for ( String key : templateArgs.keySet() ) {
+          Object value = templateArgs.get(key);
+          if ( value instanceof String ) {
+            String s = (String) value;
+            templateArgs.put(key, new String(s.getBytes(StandardCharsets.ISO_8859_1), StandardCharsets.UTF_8));
+          }
+        }
 
-JtwigTemplate templateBody =    JtwigTemplate.inlineTemplate(emailTemplate.getBody(), config);
-emailMessage.setBody(templateBody.render(model));
+        JtwigModel model = JtwigModel.newModel(templateArgs);
 
-// If subject has already provided, then we don't want to use template subject.
-if (foam.util.SafetyUtil.isEmpty(emailMessage.getSubject())) {
-  JtwigTemplate templateSubject = JtwigTemplate.inlineTemplate(emailTemplate.getSubject(), config);
-  emailMessage.setSubject(templateSubject.render(model));
-}
+        emailMessage = fillInEmailProperties(x, emailMessage, emailTemplate, model);
 
-/* If To isn't set in the message and sendTo is provided in the template, use
-  the sendTo from template. */
-String sendTo = emailTemplate.getSendTo();
-if ( emailMessage.getTo().length == 0  &&
-  ! foam.util.SafetyUtil.isEmpty(emailTemplate.getSendTo()) ) {
-  emailMessage.setTo(new String[] {sendTo});
-}
+        sendEmail(x, emailMessage);
+      `
+    },
+    {
+      name: 'fillInEmailProperties',
+      documentation: `
+      Prioritisations:
+        1) Properties set on the EmailMessage,
+        2) Properties set on the EmailTemplate,
+        3) Properties set on the Group,
+        4) Properties set as default`,
+      type: 'foam.nanos.notification.email.EmailMessage',
+      args: [
+        {
+          name: 'x',
+          type: 'Context'
+        },
+        {
+          name: 'emailMessage',
+          javaType: 'final foam.nanos.notification.email.EmailMessage'
+        },
+        {
+          name: 'emailTemplate',
+          javaType: 'final foam.nanos.notification.email.EmailTemplate'
+        },
+        {
+          name: 'model',
+          javaType: 'org.jtwig.JtwigModel'
+        }
+      ],
+      javaCode: `
+        // VARIABLE SET UP:
+        User user     = findUser(x, emailMessage);
+        Logger logger = (Logger) x.get("logger");
 
-// If the displayName doesn't set in the message
-// and the displayName provided in the template, use the displayName from template
-if ( foam.util.SafetyUtil.isEmpty(emailMessage.getDisplayName()) &&
-  ! foam.util.SafetyUtil.isEmpty(emailTemplate.getDisplayName()) ) {
-  JtwigTemplate templateDisplayName = JtwigTemplate.inlineTemplate(emailTemplate.getDisplayName(), config);
-  emailMessage.setDisplayName(templateDisplayName.render(model));
-}
+        if ( user == null || user.getId() == 0 ) {
```

Comment:
This is already caught inside of `findUser` so `user` will never be null here.

Reply:
👍